### ISSUE TITLE
Not precisely numerically correct for some update rules?

### ISSUE THREAD

**Issue by chunyang-wen** [clab/dynet#1009]
Hi. I have noticed that in `training.h`, there is a variable `sparse_updates_enabled`. In the documentation, it states that:

> they may not always be faster (particularly on GPU with mini-batch training), and are not precisely numerically correct for some update rules such as MomentumTrainer and AdamTrainer.

I'd like to know the detailed reason.

- Why slower?
- Why not numerically correct for some trainer?

Thanks in advance.

**Comment by pmichel31415** [comment]
My guess is:

- Speed: maybe checking all the gradient norms and updating only the nonzero ones sequentially is not that much faster than just updating the full lookup parameter matrix (esp. if doing minibatch training in which case there might be a lot of non-zero gradients, and on GPU where full update is faster).

- Correctness: for optimizers with momentum like MomentumSGD and Adam you need to update parameters with zero gradient if they have non-zero momentum (whereas sparse update doesn't update parameters with zero gradient regardless).

**Comment by chunyang-wen** [comment]
@pmichel31415 Thanks. By the way, I have found some related materials.

[Unorthodox Design ](https://dynet.readthedocs.io/en/latest/unorthodox.html).

I am trying to dig into different rules of updating parameters in multiple optimizers and I think it will help to understand the correctness part.
